File: python/543.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        res = 0
        
        def dfs(root: TreeNode):
            if not root: return 0
            nonlocal res
            left_depth = dfs(root.left)
            right_depth = dfs(root.right)
            res = max(res, left_depth+right_depth)
            return max(left_depth, right_depth) + 1
        
        dfs(root)
        return res

File: cpp/543.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int max_depth;
    int diameterOfBinaryTree(TreeNode* root) {
        max_depth = 0;
        depth(root);
        return max_depth;
    }
    
    int depth(TreeNode* root) {
        if (!root) return 0;
        int left_depth = depth(root->left);
        int right_depth = depth(root->right);
        max_depth = max(max_depth, left_depth + right_depth);
        return max(left_depth, right_depth)+1;
    }
};